jax/interpreters/ad.py

```diff
@@ -170,57 +170,29 @@ def write_primal(v, val):
   #        forces primal_in to contain UndefinedPrimals for tangent values!
   map(write_primal, jaxpr.invars, primals_in)
 
-  def is_linear(var):
-    if type(var) is Literal:
-      return False
-    else:
-      return var not in primal_env
-
-  linear_eqns = []
-  for eqn in jaxpr.eqns:
-    prim = eqn.primitive
-    if not (prim.call_primitive or prim.map_primitive):
-      if any(is_linear(v) for v in eqn.invars):
-        linear_eqns.append(eqn)
-      else:
-        in_vals = map(read_primal, eqn.invars)
-        ans = prim.bind(*in_vals, **eqn.params)
-        if prim.multiple_results:
-          map(write_primal, eqn.outvars, ans)
-        else:
-          write_primal(eqn.outvars[0], ans)
-    else:
-      call_jaxpr, params = core.extract_call_jaxpr(prim, eqn.params)
-      if any(is_linear(v) for v in eqn.invars):
-        linear_eqns.append(eqn)
-      if any(not is_linear(v) for v in eqn.invars):
-        # FIXME: Some invars correspond to tangents
-        ans = _eval_subjaxpr_primals(prim, call_jaxpr,
-                                     map(read_primal, eqn.invars), params)
-        map(write_primal, eqn.outvars, ans)
-
   # Find the last use of each cotangent so that they can be removed
   # as soon as possible.
   drop_cts: List[Set[Any]] = []
   seen_vars: Set[Any] = set(jaxpr.invars)
-  for eqn in linear_eqns:
+  for eqn in jaxpr.eqns:
     read_set = set(eqn.outvars)  # NOTE: eqn is not transposed yet!
     drop_cts.append(read_set - seen_vars)
     seen_vars |= read_set
 
   ct_env: Dict[Any, Any] = {}
   map(write_cotangent, jaxpr.outvars, cotangents_in)
-  for eqn, to_drop in zip(linear_eqns[::-1], drop_cts[::-1]):
+  for eqn, to_drop in zip(jaxpr.eqns[::-1], drop_cts[::-1]):
     # FIXME: Some invars correspond to tangents
     invals = map(read_primal, eqn.invars)
     if eqn.primitive.multiple_results:
       cts_in = map(read_cotangent, eqn.outvars)
     else:
       cts_in, = map(read_cotangent, eqn.outvars)
     if eqn.primitive.call_primitive or eqn.primitive.map_primitive:
+      cts_in_avals = [v.aval for v in eqn.outvars]
       call_jaxpr, params = core.extract_call_jaxpr(eqn.primitive, eqn.params)
       cts_out = get_primitive_transpose(eqn.primitive)(
-          params, call_jaxpr, invals, cts_in)
+          params, call_jaxpr, invals, cts_in, cts_in_avals)
```

Comment:
Note: an alternative to adding this argument would be to make `ad_util.zero` carry around an aval, just like `UndefinedPrimal`s do. I think we should do that as a cleanup, but I didn't want to fold it into this patch, because it's a separate thing.

Reply:
That makes sense to me!

Reply:
(As an aside: probably more of our rules would benefit from taking avals as arguments. I've noticed some things in XLA translation rules that would benefit from knowing input avals.)